File: app/test/material_read_ranges_test.dart
import 'package:auto_scoring_app/core/material_read_ranges.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('mergeMaterialRange', () {
    test('inserts into a gap without corrupting the ranges after it', () {
      // Issue #85 round 5 left a `double.nan` sentinel behind to mean "the new
      // range has been placed". Every comparison against NaN is false, so the
      // range after the gap fell through to the merge branch and a
      // `(NaN, ...)` pair reached the output.
      final merged = mergeMaterialRange(
        const [(0.0, 0.1), (0.4, 0.5), (0.8, 0.9)],
        0.2,
        0.3,
      );

      expect(merged, const [(0.0, 0.1), (0.2, 0.3), (0.4, 0.5), (0.8, 0.9)]);
      expect(
        merged.any((r) => r.$1.isNaN || r.$2.isNaN),
        isFalse,
        reason:
            'a NaN edge never compares equal to itself, so the caller '
            'would see the ranges change on every frame and rebuild forever',
      );
    });

    test('merging the same range twice changes nothing, so a redraw settles', () {
      // The defect this file exists for showed up as a redraw that never
      // stopped, not as a wrong value: the caller asks "did the ranges
      // change?" every frame, and a merge that does not converge answers
      // "yes" forever. Feeding the output back in is what states that;
      // computing the same input twice only states that the function is
      // deterministic, which every version of it was (review of PR #110).
      const cases = [
        (<(double, double)>[], 0.0, 1.0),
        ([(0.0, 0.1), (0.4, 0.5), (0.8, 0.9)], 0.2, 0.3),
        ([(0.7969, 0.9573)], 0.0764, 0.0765),
        ([(0.0, 0.2), (0.4, 0.6), (0.8, 1.0)], 0.21, 0.79),
        ([(0.2, 0.3)], 0.0, 0.1),
        ([(0.0, 0.1)], 0.9, 1.0),
      ];

      for (final (ranges, start, end) in cases) {
        final once = mergeMaterialRange(ranges, start, end);
        final twice = mergeMaterialRange(once, start, end);
        expect(
          sameMaterialRanges(once, twice),
          isTrue,
          reason:
              '$ranges + ($start, $end) does not converge -- the caller '
              'would answer "changed" on every frame and never stop rebuilding',
        );
      }
    });

    test(
      'the output is sorted, which is what lets the caller compare in order',
      () {
        final merged = mergeMaterialRange(
          const [(0.0, 0.1), (0.4, 0.5), (0.8, 0.9)],
          0.2,
          0.3,
        );

        for (var i = 1; i < merged.length; i++) {
          expect(
            merged[i - 1].$1 <= merged[i].$1,
            isTrue,
            reason:
                'sameMaterialRanges compares position by position, so an '
                'unsorted result would read as "changed" against an equal set',
          );
        }
      },
    );

    test('a rounding-sized gap still counts as touching', () {
      // What the epsilon is for. Without it these stay two ranges and the row
      // never reads as covered, so 承認 can never be pressed.
      // A literal gap, not one derived from the epsilon: deriving it means an
      // epsilon of zero closes the gap too, and the test stops discriminating
      // (found while re-running the review's mutants).
      const gap = 0.0005;
      expect(gap, lessThan(materialCoverEpsilon));
      final merged = mergeMaterialRange(const [(0.0, 0.5)], 0.5 + gap, 1.0);

      expect(merged, hasLength(1));
      expect(materialRowIsCovered(merged), isTrue);
    });

    test('absorbs ranges the new one overlaps or touches', () {
      expect(
        mergeMaterialRange(const [(0.0, 0.2), (0.3, 0.4)], 0.15, 0.35),
        const [(0.0, 0.4)],
      );
    });

    test('appends when the new range is past everything', () {
      expect(mergeMaterialRange(const [(0.0, 0.2)], 0.6, 0.7), const [
        (0.0, 0.2),
        (0.6, 0.7),
      ]);
    });

    test('places the new range first when it is before everything', () {
      expect(mergeMaterialRange(const [(0.6, 0.7)], 0.0, 0.2), const [
        (0.0, 0.2),
        (0.6, 0.7),
      ]);
    });

    test('a row is covered only once the ranges reach both edges', () {
      expect(materialRowIsCovered(null), isFalse);
      expect(materialRowIsCovered(const [(0.0, 0.5)]), isFalse);
      expect(
        materialRowIsCovered(const [(0.5, 1.0)]),
        isFalse,
        reason:
            'the top half was never on screen -- reaching the bottom edge '
            'is not the same as having seen the row',
      );
      expect(
        materialRowIsCovered(const [(0.0, 0.5), (0.5, 1.0)]),
        isFalse,
        reason: 'two ranges mean a gap was never closed',
      );
      expect(materialRowIsCovered(const [(0.0, 1.0)]), isTrue);
    });
  });

  group('sameMaterialRanges', () {
    test('looks at both edges, not just where a range starts', () {
      expect(
        sameMaterialRanges(const [(0.0, 0.5)], const [(0.0, 0.6)]),
        isFalse,
      );
      expect(
        sameMaterialRanges(const [(0.0, 0.5)], const [(0.1, 0.5)]),
        isFalse,
      );
    });

    test('a different number of ranges is a different set', () {
      expect(
        sameMaterialRanges(const [(0.0, 0.5)], const [(0.0, 0.5), (0.7, 0.9)]),
        isFalse,
      );
      expect(
        sameMaterialRanges(const [(0.0, 0.5), (0.7, 0.9)], const [(0.0, 0.5)]),
        isFalse,
      );
      expect(sameMaterialRanges(const [], const [(0.0, 0.5)]), isFalse);
    });

    test('the same set is the same', () {
      expect(
        sameMaterialRanges(const [(0.0, 0.5)], const [(0.0, 0.5)]),
        isTrue,
      );
      expect(sameMaterialRanges(const [], const []), isTrue);
    });
  });
}

File: desktop/test/material-read-ranges.test.ts
import { describe, expect, it } from "vitest";

import {
  MATERIAL_COVER_EPSILON,
  materialRowGapStart,
  materialRowIsCovered,
  mergeMaterialRange,
  planMaterialReveal,
  sameMaterialRanges,
} from "../src/renderer/core/material-read-ranges.js";

describe("mergeMaterialRange (INV-065, INV-066)", () => {
  it("INV-065: inserts into a gap without NaN corruption", () => {
    const merged = mergeMaterialRange(
      [
        [0, 0.1],
        [0.4, 0.5],
        [0.8, 0.9],
      ],
      0.2,
      0.3,
    );
    expect(merged).toEqual([
      [0, 0.1],
      [0.2, 0.3],
      [0.4, 0.5],
      [0.8, 0.9],
    ]);
    expect(merged.some((r) => Number.isNaN(r[0]) || Number.isNaN(r[1]))).toBe(
      false,
    );
  });

  it("converges when merging the same range twice", () => {
    const cases: [readonly [number, number][], number, number][] = [
      [[], 0, 1],
      [
        [
          [0, 0.1],
          [0.4, 0.5],
          [0.8, 0.9],
        ],
        0.2,
        0.3,
      ],
      [[[0.7969, 0.9573]], 0.0764, 0.0765],
    ];
    for (const [ranges, start, end] of cases) {
      const once = mergeMaterialRange(ranges, start, end);
      const twice = mergeMaterialRange(once, start, end);
      expect(sameMaterialRanges(once, twice)).toBe(true);
    }
  });

  it("INV-066: a row is covered only once both edges are reached", () => {
    expect(materialRowIsCovered(null)).toBe(false);
    expect(materialRowIsCovered([[0, 0.5]])).toBe(false);
    expect(materialRowIsCovered([[0.5, 1]])).toBe(false);
    expect(
      materialRowIsCovered([
        [0, 0.5],
        [0.5, 1],
      ]),
    ).toBe(false);
    expect(materialRowIsCovered([[0, 1]])).toBe(true);
  });

  it("rounding-sized gap still counts as touching", () => {
    const gap = 0.0005;
    expect(gap).toBeLessThan(MATERIAL_COVER_EPSILON);
    const merged = mergeMaterialRange([[0, 0.5]], 0.5 + gap, 1);
    expect(merged).toHaveLength(1);
    expect(materialRowIsCovered(merged)).toBe(true);
  });

  it("INV-319: reports where the first unread part of a row begins", () => {
    expect(materialRowGapStart(null)).toBe(0);
    expect(materialRowGapStart([])).toBe(0);
    expect(materialRowGapStart([[0, 1]])).toBeNull();
    // Seen from the top only: the gap is below the seen part.
    expect(materialRowGapStart([[0, 0.6]])).toBe(0.6);
    // Seen from the bottom only: the gap is above.
    expect(materialRowGapStart([[0.4, 1]])).toBe(0);
    // A middle gap after a first pass.
    expect(
      materialRowGapStart([
        [0, 0.3],
        [0.6, 1],
      ]),
    ).toBe(0.3);
  });
});

describe("planMaterialReveal (Issue #328)", () => {
  const base = {
    maxScrollTop: 1616,
    viewportHeight: 608,
    containerTop: 0,
  };

  it("brings the first unread point just inside the top edge", () => {
    // One pixel inside, not exactly on the edge: the browser rounds the scroll
    // offset, and landing on the edge left a sub-pixel sliver the gate refused
    // (measured 0.56px on the real container).
    expect(
      planMaterialReveal({ ...base, scrollTop: 0, unreadTop: 1160 }),
    ).toEqual({ scrollTop: 1159, moved: true });
  });

  it("moves upward when the unread point is above the viewport", () => {
    expect(
      planMaterialReveal({ ...base, scrollTop: 600, unreadTop: -200 }),
    ).toEqual({ scrollTop: 399, moved: true });
  });

  it("clamps the unread point to the reachable scroll range", () => {
    expect(
      planMaterialReveal({ ...base, scrollTop: 1600, unreadTop: 500 }),
    ).toEqual({ scrollTop: 1616, moved: true });
  });

  it("still advances when the unread point is already at the top", () => {
    // unreadTop 1 cancels the edge margin, so the alignment is a no-op and the
    // one-viewport fallback takes over instead of stalling.
    expect(
      planMaterialReveal({
        maxScrollTop: 600,
        viewportHeight: 300,
        containerTop: 0,
        scrollTop: 300,
        unreadTop: 1,
      }),
    ).toEqual({ scrollTop: 600, moved: true });
  });

  it("reports no movement when the container has no scroll range", () => {
    // Without this the button would claim to have revealed material it never
    // moved (Issue #328, decision 2).
    expect(
      planMaterialReveal({
        maxScrollTop: 0,
        viewportHeight: 608,
        containerTop: 0,
        scrollTop: 0,
        unreadTop: 608,
      }),
    ).toEqual({ scrollTop: 0, moved: false });
  });
});
